src/Http/Http.Extensions/test/RequestDelegateGenerator/RequestDelegateGeneratorTests.cs

```diff
@@ -516,6 +517,88 @@ public async Task MapAction_UnknownParameter_EmitsDiagnostic_NoSource()
+            var fromServiceEnumerableDefaultValueSource = $$"""
+#nullable disable
+string postServiceWithDefault([FromServices]IEnumerable<{{typeof(TestService)}}> svc = null) => svc?.FirstOrDefault()?.TestServiceMethod() ?? string.Empty;
+app.MapPost("/", postServiceWithDefault);
+#nullable restore
+""";
+
+            return new[]
+            {
+                new object[] { fromServiceRequiredSource, true, true },
+                new object[] { fromServiceRequiredSource, false, false },
+                new object[] { fromServiceNullableSource, true, true },
+                new object[] { fromServiceNullableSource, false, true },
+                new object[] { fromServiceDefaultValueSource, true, true },
+                new object[] { fromServiceDefaultValueSource, false, true },
+                new object[] { fromServiceEnumerableRequiredSource, true, true },
+                new object[] { fromServiceEnumerableRequiredSource, false, true },
+                new object[] { fromServiceEnumerableNullableSource, true, true },
+                new object[] { fromServiceEnumerableNullableSource, false, true },
+                new object[] { fromServiceEnumerableDefaultValueSource, true, true },
+                new object[] { fromServiceEnumerableDefaultValueSource, false, true }
+            };
+        }
+    }
+
+    [Theory]
+    [MemberData(nameof(MapAction_ExplicitServiceParam_SimpleReturn_Data))]
+    public async Task MapAction_ExplicitServiceParam_SimpleReturn(string source, bool hasService, bool isValid)
+    {
+        var (_, compilation) = await RunGeneratorAsync(source);
+        var endpoint = GetEndpointFromCompilation(compilation);
+
+        var httpContext = CreateHttpContext();
+        if (hasService)
+        {
+            var serviceCollection = new ServiceCollection();
+            serviceCollection.AddSingleton<TestService>(new TestService());
+            var services = serviceCollection.BuildServiceProvider();
+            httpContext.RequestServices = services;
+        }
+
+        if (isValid)
+        {
+            await endpoint.RequestDelegate(httpContext);
+            await VerifyResponseBodyAsync(httpContext, hasService ? "Produced from service!" : string.Empty);
+        }
+        else
+        {
+            await Assert.ThrowsAsync<InvalidOperationException>(() => endpoint.RequestDelegate(httpContext));
+            Assert.False(httpContext.RequestAborted.IsCancellationRequested);
+        }
+    }
+
+    [Fact]
+    public async Task MapAction_ExplicitServiceParam_SimpleReturn_Snapshot()
+    {
+        var source = $$"""
+app.MapPost("/fromServiceRequired", ([FromServices]{{typeof(TestService)}} svc) => svc.TestServiceMethod());
+app.MapPost("/enumerableFromService", ([FromServices]IEnumerable<{{typeof(TestService)}}> svc) => svc?.FirstOrDefault()?.TestServiceMethod() ?? string.Empty);
```

Comment:
Gotta love when you get arrays of service types for free from the service provider ;)